components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -336,189 +577,247 @@ impl RangeManager {
-            return Err(LoadFailedReason::InGc);
+    pub fn load_region(&mut self, region: Region) -> Result<(), LoadFailedReason> {
+        use RegionState::*;
+        if let Some(state) = self.check_overlap_with_region(&region) {
+            let reason = match state {
+                Pending | ReadyToLoad | Loading => LoadFailedReason::PendingRange,
+                Active => LoadFailedReason::Overlapped,
+                LoadingCanceled | PendingEvict | Evicting => LoadFailedReason::Evicting,
+            };
+            return Err(reason);
         }
-        if self.overlap_with_evicting_range(&cache_range) {
-            return Err(LoadFailedReason::Evicting);
-        }
-        self.pending_ranges.push(cache_range);
+        let meta = RangeMeta::new(region);
+        self.new_region_meta(meta);
         Ok(())
     }
 
-    // Only ranges that have not been scheduled will be retained in `ranges`
-    pub fn mark_delete_ranges_scheduled(&mut self, ranges: &mut Vec<CacheRange>) {
-        ranges.retain(|r| {
-            let (ref mut scheduled, ..) = self.ranges_being_deleted.get_mut(r).unwrap();
-            let has_scheduled = *scheduled;
-            *scheduled = true;
-            !has_scheduled
-        });
+    // return `true` is the region is evicted.
+    pub(crate) fn split_region(
+        &mut self,
+        source_region: &Region,
+        mut new_regions: Vec<kvproto::metapb::Region>,
+    ) {
+        if let Some(region_meta) = self.region_meta(source_region.id) {
+            // if region is evicting, skip handling split for simplicity.
+            if region_meta.state >= RegionState::LoadingCanceled {
+                info!("region is evicted, skip split"; "meta" => ?&region_meta, "new_regions" => ?new_regions);
+                return;
+            }
+        } else {
+            info!("split source region not cached"; "region_id" => source_region.id);
+            return;
+        }
+
+        let region_meta = self.remove_region(source_region.id);
+        assert!(region_meta.state <= RegionState::Active);
+        if region_meta.region.get_region_epoch().version != source_region.get_region_epoch().version
+        {
+            // for pending regions, we keep regions that still fall in this range if epoch
+            // version changed.
+            assert_eq!(region_meta.state, RegionState::Pending);
+            assert!(
+                region_meta.region.get_region_epoch().version
+                    < source_region.get_region_epoch().version
+            );
+            new_regions.retain(|r| {
+                r.start_key <= source_region.start_key
+                    && (source_region.end_key.is_empty()
+                        || (source_region.end_key >= r.end_key && !r.end_key.as_slice().is_empty()))
+            });
+            info!("[IME] handle split region met pending region epoch stale"; "cached" => ?region_meta, "split_source" => ?source_region, "cache_new_regions" => ?new_regions);
```

Comment:
Please keep line width within 100 chars (80 even better).

Reply:
Is it possible the let cargo format due this for us automatically?🤔

Reply:
rustfmt does not work well for complex macros.